cesium/internal/domain/index.go

```diff
@@ -49,14 +54,33 @@ func (idx *index) insert(ctx context.Context, p pointer) error {
 		} else if !idx.beforeFirst(p.End) {
 			i, overlap := idx.unprotectedSearch(p.TimeRange)
 			if overlap {
-				idx.mu.RUnlock()
-				return span.EndWith(ErrDomainOverlap)
+				idx.mu.Unlock()
+				return span.Error(ErrDomainOverlap)
 			}
 			insertAt = i + 1
 		}
 	}
-	idx.mu.RUnlock()
-	idx.insertAt(ctx, insertAt, p)
+
+	if insertAt == 0 {
+		idx.mu.pointers = append([]pointer{p}, idx.mu.pointers...)
+	} else if insertAt == len(idx.mu.pointers) {
+		idx.mu.pointers = append(idx.mu.pointers, p)
+	} else {
+		idx.mu.pointers = append(idx.mu.pointers[:insertAt], append([]pointer{p}, idx.mu.pointers[insertAt:]...)...)
+	}
+
+	idx.persistHead = min(idx.persistHead, insertAt)
+
+	if persist {
+		update := indexUpdate{afterIndex: idx.persistHead}
+		idx.persistHead = len(idx.mu.pointers) - 1
+		idx.mu.Unlock()
+		idx.Observer.Notify(ctx, update)
```

Comment:
Honestly we should probably just get rid of the observer here and call the persist function directly. Adding an observer just adds complexity and was kinda dumb on my part.

Reply:
Agreed – also allows us to get rid of the unlocking - relocking semantic. Was there any advantage in having an observer? I guess the index persist would run concurrently with other db operations so faster?